devdoc/channel.md

```diff
@@ -0,0 +1,294 @@
+
+`channel_pull` and `channel_push` can be called from different threads in any order. The callbacks are executed on threadpool threads.
+
+## `channel` internal state
+
+The internal state of the `channel` module can have the following values:
+  - `CHANNEL_EMPTY`: This is the initial state of the `channel`. It means that there is no data available to be pulled. Calling `channel_pull` or `channel_push` will cause the state to transition to `CHANNEL_PUSHED` or `CHANNEL_PULLED` respectively. After all the `channel_pull` and `channel_push` calls have been paired, the state will transition back to `CHANNEL_EMPTY`.
+  - `CHANNEL_PUSHED`: This is the state after `channel_push` has been called. A subsequent call to `channel_pull` will schedule the callbacks on the threadpool.
+  - `CHANNEL_PULLED`: This is the state after `channel_pull` has been called. A subsequent call to `channel_push` will the callbacks on the threadpool.
+
+```mermaid
+stateDiagram-v2
+    B: STATE=CHANNEL_EMPTY
+    N: STATE=CHANNEL_PUSHED
+    R: STATE=CHANNEL_PULLED
+    E: STATE=CHANNEL_EMPTY
+
+    [*] --> B: channel_create
+
+    B --> N: channel_push
+    B --> R: channel_pull
+
+    N --> E: channel_pull
+
+    R --> E: channel_push
+
+    N --> E: cancel channel_push
+    R --> E: cancel channel_pull
+
+    R --> R: channel_pull
+    N --> N: channel_push
+
+    E --> [*]: channel_destroy
+```
+
+## Operations
+
+The first call to `channel_pull`/`channel_push` either creates an `operation` and adds it to a `list of pending operations`. When `channel_push`/`channel_pull` is called subsequently, the least-recently added `operation` in the `list of pending operations` is scheduled to have its callbacks called on the threadpool. Each operation maintains an internal `state`. The `state` of each `operation` can have the following values:
+
+ - `CHANNEL_OP_CREATED`: This is the initial state of the `operation` after only one of `channel_pull`/`channel_push` have been called. It means that the `operation` has been created and added to the `list of pending operations`. A subsequent call to `channel_push`/`channel_pull` will cause the `operation` to transition to `CHANNEL_OP_SCHEDULED`.
+
+ - `CHANNEL_OP_SCHEDULED`: This is the state after the a pair of `channel_pull`/`channel_push` have been called. It means that the `operation` has been removed from the `list of pending operations` and is scheduled to have its callbacks called on the threadpool with result `CHANNEL_CALLBACK_RESULT_OK`. 
+
+ - `CHANNEL_OP_CANCELLED`: This is the state after the `operation` has been cancelled. It means that the `operation` has been removed from the `list of pending operations` and is scheduled to have a callback (either `PULL_CALLBACK` XOR `PUSH_CALLBACK`, depending on which of `channel_pull`/`channel_push` created it) called with result `CHANNEL_CALLBACK_RESULT_CANCELLED`. 
```

Comment:
>XOR [](http://example.com/codeflow?start=232&length=3)

Why XOR? I believe we've discussed that operations can be cancelled even after both the callbacks exist and the operation has been transferred to the threadpool.

Reason to wait, because either me doesn't understand what XOR means or this is not what we have discussed yesterday.

I believe the "to have **a** callback" means only one which sort of reinforces that I understand what XOR means in this context AND this is not what I expect the specs to say.

Reply:
Removing XOR